### DISCUSSION TITLE
Next 13.4.1 : Images, Js and Css not loading after build. Shows 404 for all

### DISCUSSION THREAD

**Discussion by Ashikpaul** [vercel/next.js#49803] [Help] [answered]
### Summary

I'm trying to deploy my next 13 app. I don't see any errors while building. When I try to load the app, only HTML shows up, and it shows 404 error for Js, Css and images(as seen in the network tab).

Its trying to load from _next folder by default (and this path does not exist)
Eg: http://localhost:3000/_next/static/chunks/522-b7eb9fb4e38abfae.js

And images directly
Eg: http://localhost:3000/assets/login.png

Unfortunately I can't post the src code.
I have tried everything. Any help is appreciated.

### Additional information

```js
Folder structure

My-app
|_ public
  |_ assets/
    |_[all my images]
|_ nodemodules
|_ .next
|_ out
|_ src
  |_ app
|_ Dockerfile
|_ next.config.js
|_ package.json
```

Below are the scripts using

next.config.js
```
module.exports = {
  basePath: "",
  output: "standalone",
  experimental: {
    appDir: true,
  },
  images: {
    domains: [
      "localhost",
      "localhost:3001",
      "my-exapmle-domain.com",
    ],
  },
};
```

dockerfile
```
#Using a different image, below is just an example
FROM node:16_alpine-3.16 AS prebuild

USER root

WORKDIR /app

ARG env 
ENV NODE_ENV=${env}

RUN apk add --no-cache libc6-compat
RUN apk add --update npm

COPY package.json ./
COPY package-lock.json ./
RUN npm i --legacy-peer-deps

COPY . .

RUN npm run build

#Using a different image, below is just an example
FROM node:16_alpine-3.16

USER root

WORKDIR /app

ENV NEXT_TELEMETRY_DISABLED 1
ENV UID=10101
ENV GID=10101

## create & switch to non-root user
RUN apk add shadow
RUN /usr/sbin/groupadd -g ${GID} docker
RUN /usr/sbin/useradd -s /bin/sh -g ${GID} -u ${UID} limsuinext

USER 10101

COPY --from=prebuild /app/next.config.js ./
COPY --from=prebuild /app/package.json ./package.json
COPY --from=prebuild /app/.next ./.next
COPY --from=prebuild /app/node_modules ./node_modules
COPY --from=prebuild /app/public/ ./public


COPY --from=prebuild --chown=10101 /app/.next/standalone ./
COPY --from=prebuild --chown=10101 /app/.next/static ./.next/static
COPY --from=prebuild --chown=10101 /app/.next/static ./.next/standalone/static
COPY --from=prebuild --chown=10101 /app/.next/out ./.next/standalone/out


# Fire up node server
ENV PORT 3001
EXPOSE 3001
# CMD ["node", ".next/standalone/server.js"]  Old approach
CMD ["node", "server.js"]
```

package.json
```
name: "my-app",
version: "0.1.0",
private: true,
scripts: {
    "dev": "PORT=3001 npx next dev",
    "build": "npx next build",
    "start": "PORT=3001 next start",
    "lint": "next lint"
},
dependencies: {...},
devDependencies: {...}
```

**Comment by icyJoseph** [comment]
How do you start the server?

**Reply by Ashikpaul** [reply]
```
docker build -t client-ui-next -f Dockerfile . && docker run -p 3001:3001 client-ui-next
```
Running the above command locally loads everything fine

But when i try the below commands locally it fails (which happens same as the production env, exact behaviour)
```
npm run build
```
then
```
node .next/standalone/server.js 
```

**Reply by icyJoseph** [reply]
> Additionally, a minimal server.js file is also output which can be used instead of next start. **_This minimal server does not copy the public or .next/static folders by default as these should ideally be handled by a CDN instead, although these folders can be copied to the standalone/public and standalone/.next/static folders manually, after which server.js file will serve these automatically._**

From: https://nextjs.org/docs/app/api-reference/next-config-js/output

So you have to manually copy public and .next/static folders.

**Reply by Ashikpaul** [reply]
Had tried this, which didn't work. The folders are not visible / getting copied.
"_next" folder is not even there. No matter what i try.

```
#
# Prepare Pre Build  Artifact
#

FROM e-isgcbi.docker.lowes.com/library/isg-image-node:16_alpine-3.16 AS prebuild

USER root

WORKDIR /app

ARG lims_asset_path
ENV LIMS_ASSET_PATH=${lims_asset_path}

ARG lims_base_name
ENV LIMS_BASE_NAME=${lims_base_name}

ARG env 
ENV NODE_ENV=${env}

RUN apk add --no-cache libc6-compat
RUN apk add --update npm

COPY package.json ./
COPY package-lock.json ./
RUN npm i --legacy-peer-deps

COPY . .

RUN npm run build

#
# Prepare Final Deployable Artifact
#

FROM e-isgcbi.docker.lowes.com/library/isg-image-node:16_alpine-3.16

USER root

WORKDIR /app

ENV NEXT_TELEMETRY_DISABLED 1
ENV UID=10101
ENV GID=10101

## create & switch to non-root user
RUN apk add shadow
RUN /usr/sbin/groupadd -g ${GID} docker
RUN /usr/sbin/useradd -s /bin/sh -g ${GID} -u ${UID} limsuinext

USER 10101

# COPY --from=prebuild --chown=10101 /app/.next/standalone ./standalone
# COPY --from=prebuild --chown=10101 /app/public /app/standalone/public
# COPY --from=prebuild --chown=10101 /app/.next/static /app/standalone/.next/static
# COPY --from=prebuild --chown=10101 /app/.next/cache /app/standalone/.next/cache
# COPY --from=prebuild --chown=10101 /app/.next/server ./.next/server
# COPY --from=prebuild --chown=10101 /app/public ./public

COPY --from=prebuild /app/next.config.js ./
COPY --from=prebuild /app/package.json ./package.json
COPY --from=prebuild /app/.next ./.next
COPY --from=prebuild /app/node_modules ./node_modules
COPY --from=prebuild /app/public ./public


COPY --from=prebuild --chown=10101 /app/.next/standalone ./
COPY --from=prebuild --chown=10101 /app/.next/static ./.next/static
COPY --from=prebuild --chown=10101 /app/.next/static ./.next/standalone/.next/static
COPY --from=prebuild --chown=10101 /app/public ./.next/standalone/public


# Fire up node server
ENV PORT 3001
EXPOSE 3001
CMD ["node", "server.js"]
```

**Reply by icyJoseph** [reply]
but I don't understand, are you building/running using docker, or not 🤔

**Reply by Ashikpaul** [reply]
Yes, I am building and running using docker.

**Reply by Ashikpaul** [reply]
@icyJoseph  This happens 

<img width="1094" alt="Screenshot 2023-05-15 at 6 23 55 PM" src="https://github.com/vercel/next.js/assets/17526871/560bfe11-f228-4c74-bb1a-4475b7e38a1e">

**Reply by icyJoseph** [reply]
I must be really misunderstanding you. When you deploy, are you using Docker? Or are you just running manually building and starting the node server?

**Reply by Ashikpaul** [reply]
Locally I don't need these but the server in prod uses the dockerfile (which is failing). Hence I am trying to mock this failure using the commands I mentioned in the below link.

https://github.com/vercel/next.js/discussions/49803#discussioncomment-5904279

The reason why its showing 404 error in the network tab is because that path doesn't exist (this is what i think, may be i am wrong). I couldn't find a similar settings in anybody's setup. I don't know what to do.

I read that next automatically created ```_next``` folder. If this was true all the files should have been there, which i don't see it.

My only goal is to run Next 13 setup using Dockerfile (prod setup). Maybe my approach is wrong.

**Reply by icyJoseph** [reply]
I understand why the 404's but I just couldn't follow how you deploy the app.

Ok, so the production environment builds the docker file and runs the resulting image. And even though it does that, the static files are unreachable. Correct?

**Reply by Ashikpaul** [reply]
Exactly. Sorry I didn't make it clear

**Reply by icyJoseph** [reply]
Just one more question, the screenshot you showed, why does it say localhost? Is it a screenshot taken locally?

**Reply by Ashikpaul** [reply]
Yes. Its the same error in prod with a different host

**Comment by Ashikpaul** [comment, marked as answer]
The Fix is
Manually copy the _next folder in dockerfile
For js and css
```COPY --from=builder --chown=10101 /app/.next ./_next```
For images
```COPY --from=builder --chown=10101 /app/public ./public```

Thank you @icyJoseph

**Comment by khal3d** [comment]
@Ashikpaul I have a similar issue. Any luck figuring out what's causing it?

**Reply by abhinavgurung** [reply]
@khal3d  Hello I have similar issue. Mine is deployed in aks. did you find the solution?

**Comment by Ashikpaul** [comment]
@khal3d, @abhinavgurung 

```
FROM PATH_TO_YOUR_NODE_IMAGE AS prebuild

USER root

WORKDIR /app

ARG env 
ENV NODE_ENV=dev

RUN apk add --no-cache libc6-compat
RUN apk add --update npm

COPY ./MY_PROJECT .


RUN npm i --legacy-peer-deps

WORKDIR /app

RUN npm run build

/*Prepare Final Deployable Artifact*/

FROM PATH_TO_YOUR_NODE_IMAGE 

USER root

WORKDIR /app

ENV NEXT_TELEMETRY_DISABLED 1
ENV UID=10101
ENV GID=10101

/* create & switch to non-root user */
RUN apk add shadow
RUN /usr/sbin/groupadd -g ${GID} docker
RUN /usr/sbin/useradd -s /bin/sh -g ${GID} -u ${UID} MY_PROJECT

USER 10101

COPY --from=prebuild --chown=10101 /app/.next/standalone ./
COPY --from=prebuild --chown=10101 /app/.next/static ./.next/static
COPY --from=prebuild --chown=10101 /app/.next/static ./.next/standalone/.next/static
COPY --from=prebuild --chown=10101 /app/public ./
COPY --from=prebuild --chown=10101 /app/public ./public
COPY --from=prebuild --chown=10101 /app/.next/server ./.next/standalone/server
COPY --from=prebuild --chown=10101 /app/.next/cache ./.next/standalone/cache
COPY --from=prebuild --chown=10101 /app/.next/static ./_next/static
COPY --from=prebuild --chown=10101 /app/.next ./_next


/* Fire up node server */
ENV PORT 3000
EXPOSE 3000
CMD ["node", "server.js"]
```

This is the one that works for me.
Hope this helps

**Comment by dbain-texada** [comment]
I'm wondering how this changes at all if I am setting basePath: '/docs', for instance in my next.config.mjs file? Seems like everything works as expected running npm run dev (next dev -p 3002), however, after building DockerFile image it seems the basePath is ignored and I can only access localhost:3002 and all my static files http://localhost:3002/docs/_next/static/chunks/XXXXXXXXX etc. return 404?

**Reply by plamabo** [reply]
hey dbain, that's interesting we are at the same point now a week later. Did you have any additional insights? 
thanks :)
matthias

**Comment by gilbertlucas46** [comment]
I had update next config to

const nextConfig = {
  distDir: '_next',
}

locally it works fine but when I deployed in production the build files is in _next folder instead of .next
